### PR TITLE
[rllib] First pass at pipeline implementation of DQN

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds a bare bones implementation of DQN with the pipelines impl. Prioritized replay / Multi-agent / Ape-X support are left for a future PR.

It can be compared against the existing DQN with:
```
./train.py --run=DQN --env=CartPole-v0 --config='{"use_pipeline_impl": {"grid_search": [1, 0]}}'
```

### CODE REVIEW

File: rllib/agents/dqn/dqn.py [hunk 1]
```diff
@@ -308,6 +312,25 @@ def update_target_if_needed(trainer, fetches):
         trainer.state["num_target_updates"] += 1
 
 
+def training_pipeline(workers, config):
+    local_replay_buffer = ReplayBuffer(config["buffer_size"])
+    rollouts = ParallelRollouts(workers)
```
Comment (sven1977) [inline rllib/agents/dqn/dqn.py]: Can you explicitly set the `mode` here for clarity?

File: rllib/agents/dqn/dqn.py [hunk 2]
```diff
@@ -308,6 +312,25 @@ def update_target_if_needed(trainer, fetches):
...
+
+    # (1) Store experiences into the local replay buffer.
+    store_op = rollouts.for_each(StoreToReplayBuffer(local_replay_buffer))
+
+    # (2) Replay and train on selected experiences.
+    replay_op = LocalReplay(local_replay_buffer, config["train_batch_size"]) \
```
Comment (sven1977) [inline rllib/agents/dqn/dqn.py]: It's not intuitively clear here, why we have to call `for_each` on the return value of LocalReplay (intuitively, I would think it's a single batch taken by the learner from the (only) replay_buffer). Could you add more comments here to explain?

File: rllib/evaluation/worker_set.py [hunk 3]
```diff
@@ -71,6 +72,12 @@ def remote_workers(self):
         """Return a list of remote rollout workers."""
         return self._remote_workers
 
+    def sync_weights(self):
```
Comment (sven1977) [inline rllib/evaluation/worker_set.py]: Can we add a tau-option here for soft synching?

File: rllib/utils/experimental_dsl.py [hunk 4]
```diff
@@ -434,3 +472,125 @@ def __call__(self, gradients):
...
+                    "weights": weights,
+                    "batch_indexes": batch_indexes
+                })
+            yield MultiAgentBatch(samples, train_batch_size)
+
+    return LocalIterator(gen_replay, MetricsContext())
+
+
+def Concurrently(ops: List[LocalIterator], mode="round_robin"):
+    """Operator that run the given parent iterators concurrently.
```
Comment (sven1977) [inline rllib/utils/experimental_dsl.py]: "runs"

**Review by sven1977 — CHANGES_REQUESTED** [review]

File: rllib/evaluation/worker_set.py (hunk 3 above)
Comment (ericl) [reply rllib/evaluation/worker_set.py]: This is just a convenience method for existing hard sync. Soft sync should be done at the local level not between workers.

**Review by sven1977 — APPROVED** [review]

**Pull request merged**